android/src/main/java/com/reactnativehmssdk/HMSRNSDK.kt

```diff
@@ -1896,4 +1917,118 @@ class HMSRNSDK(
+      val key = data.getString("key")!!
+      val value = data.getString("value")
+
+      sessionStore.set(
+        value, // data/value
+        key, // key
+        object : HMSActionResultListener {
+          override fun onError(error: HMSException) {
+            promise?.reject(error.code.toString(), error.message)
+          }
+          override fun onSuccess() {
+            val result: WritableMap = Arguments.createMap()
+            result.putBoolean("success", true)
+            result.putString("finalValue", value)
+            promise?.resolve(result)
+          }
+        }
+      )
+    } else {
+      val errorMessage = "setSessionMetadataForKey: $requiredKeys"
+      rejectCallback(promise, errorMessage)
+    }
+  }
+
+  fun getSessionMetadataForKey(data: ReadableMap, promise: Promise?) {
+    val requiredKeys = HMSHelper.getUnavailableRequiredKey(data, arrayOf(Pair("key", "String")))
+    if (requiredKeys === null) {
+      val key = data.getString("key")!!
+
+      sessionStore.get(key, object : HMSSessionMetadataListener {
+        override fun onError(error: HMSException) {
+          promise?.reject(error.code.toString(), error.message)
+        }
+        override fun onSuccess(sessionMetadata: Any?) {
+          if (sessionMetadata is String?) {
+            promise?.resolve(sessionMetadata)
+          } else {
+            promise?.reject("6002", "Session Store: Unsupported type received for '$key' key, only String type is supported")
+          }
+        }
+      })
+    } else {
+      val errorMessage = "getSessionMetadataForKey: $requiredKeys"
+      rejectCallback(promise, errorMessage)
+    }
+  }
+
+  fun addKeyChangeListener(data: ReadableMap, promise: Promise?) {
+    val requiredKeys = HMSHelper.getUnavailableRequiredKey(data, arrayOf(Pair("keys", "Array"), Pair("uniqueId", "String")))
+    if (requiredKeys === null) {
+      val keys = ArrayList(data.getArray("keys")!!.toArrayList().map { it.toString() })
+      val uniqueId = data.getString("uniqueId")!!
+
+      val keyChangeListener = object : HMSKeyChangeListener {
+        override fun onKeyChanged(key: String, value: Any?) {
+          val map = Arguments.createMap()
+          map.putString("id", id)
+          map.putString("key", key)
+          if (value is String?) {
+            map.putString("value", value)
+          } else {
+            Log.i("HMSRNSDK", "Session Store: '$value' value received for '$key' key, expected only NullableString type for value")
+            map.putString("value", null) // resetting value to `null`, as the current type is not supported
+          }
+          delegate.emitEvent(uniqueId, map)
+        }
+      }
+
+      val actionResultListener = object : HMSActionResultListener {
+        override fun onError(error: HMSException) {
+          promise?.reject(error.code.toString(), error.message)
+        }
+        override fun onSuccess() {
+          keyChangeObservers[uniqueId] = keyChangeListener
```

Comment:
the list of `keyChangeObservers` should also be cleared when Room is left, or Peer is removed from Room. Since this var persists, Cleanup actions should be performed